File: Golang/Easy/3432_count_partitions_with_even_sum_difference.go
package main

import "fmt"

// countPartitions returns the number of partitions with an even sum difference.
//
// For any partition, diff = 2L - S where L is the left sum and S the total.
// Since 2L is always even, the parity of diff depends only on S:
//   - S even → all n-1 partitions are valid
//   - S odd  → no partition is valid
//
// Time complexity:  O(n)
// Space complexity: O(1)
func countPartitions(nums []int) int {
	sum := 0
	for _, val := range nums {
		sum += val
	}
	if sum%2 == 0 {
		return len(nums) - 1
	}
	return 0
}

// countPartitionsLoop returns the count by maintaining a running left sum,
// computing the difference at each partition index without re-summing.
//
// Time complexity:  O(n)
// Space complexity: O(1)
func countPartitionsLoop(nums []int) int {
	total := 0
	for _, num := range nums {
		total += num
	}
	counter := 0
	sumLeft := 0
	for i, x := range nums {
		if i == len(nums)-1 {
			continue
		}
		sumLeft += x
		diff := (total - sumLeft) - sumLeft
		if diff%2 == 0 {
			counter++
		}
	}
	return counter
}

func main() {
	testCases := []struct {
		nums     []int
		expected int
	}{
		{[]int{10, 10, 3, 7, 6}, 4},
		{[]int{1, 2, 2}, 0},
		{[]int{2, 4, 6, 8}, 3},
		{[]int{1, 1}, 1},
		{[]int{1, 2}, 0},
	}

	for i, tc := range testCases {
		r1, r2 := countPartitions(tc.nums), countPartitionsLoop(tc.nums)
		s1, s2 := "PASS", "PASS"
		if r1 != tc.expected {
			s1 = "FAIL"
		}
		if r2 != tc.expected {
			s2 = "FAIL"
		}
		fmt.Printf("Example %d | Expected: %d | parity: %d [%s] | loop: %d [%s]\n",
			i+1, tc.expected, r1, s1, r2, s2)
	}
}

File: Python/Easy/3432_count_partitions_with_even_sum_difference.py
"""
3432. Count Partitions with Even Sum Difference

You are given an integer array nums of length n.

A partition is defined as an index i where 0 <= i < n - 1, splitting the
array into two non-empty subarrays such that:
  Left subarray contains indices [0, i].
  Right subarray contains indices [i + 1, n - 1].

Return the number of partitions where the difference between the sum of the
left and right subarrays is even.

Constraints:
- 2 <= n == nums.length <= 100
- 1 <= nums[i] <= 100

Examples:

Example 1:
    Input: nums = [10,10,3,7,6]
    Output: 4

Example 2:
    Input: nums = [1,2,2]
    Output: 0

Example 3:
    Input: nums = [2,4,6,8]
    Output: 3
"""


class Solution:
    def countPartitions(self, nums: list[int]) -> int:
        """
        Returns the count of valid partitions using a parity argument.

        For any partition, let L = left sum and S = total sum. Then:
            diff = L - R = L - (S - L) = 2L - S

        Since 2L is always even, the parity of diff depends solely on S:
          - S even  →  2L - S is always even  →  all n-1 partitions are valid
          - S odd   →  2L - S is always odd   →  no partition is valid

        Time complexity:  O(n) to compute the sum
        Space complexity: O(1)
        """
        return len(nums) - 1 if sum(nums) % 2 == 0 else 0

    def countPartitionsLoop(self, nums: list[int]) -> int:
        """
        Returns the count of valid partitions by brute-force enumeration.

        The early exit on odd total avoids the O(n²) loop entirely, since no
        partition can qualify. When the total is even every partition qualifies,
        but the loop still demonstrates the full check.

        Time complexity:  O(n²) due to repeated sum() calls inside the loop
        Space complexity: O(n) for each slice
        """
        if sum(nums) % 2 != 0:
            return 0
        ans: int = 0
        for i in range(len(nums) - 1):
            diff: int = sum(nums[:i]) - sum(nums[i:])
            if diff % 2 == 0:
                ans += 1
        return ans


if __name__ == "__main__":
    solution = Solution()

    test_cases = [
        ([10, 10, 3, 7, 6], 4),
        ([1, 2, 2], 0),
        ([2, 4, 6, 8], 3),
        ([1, 1], 1),
        ([1, 2], 0),
    ]

    for label, method in [
        ("countPartitions    ", solution.countPartitions),
        ("countPartitionsLoop", solution.countPartitionsLoop),
    ]:
        for nums, expected in test_cases:
            result = method(nums)
            assert result == expected, \
                f"{label} failed: nums={nums} → expected {expected}, got {result}"
        print(f"  {label} passed")

    print("All test cases passed!")
